client/requester.go

```diff
@@ -0,0 +1,292 @@
+}
+
+// rawOnce is unchanged from the client implementation
+func (br *BaseRequester) raw(ctx context.Context, method, urlpath string, query url.Values, headers map[string]string, body io.Reader) (*http.Response, error) {
+	// fake a url to keep http.Client happy
+	u := br.baseURL
+	u.Path = path.Join(br.baseURL.Path, urlpath)
+	u.RawQuery = query.Encode()
+	req, err := http.NewRequestWithContext(ctx, method, u.String(), body)
+	if err != nil {
+		return nil, RequestError{err}
+	}
+	if br.userAgent != "" {
+		req.Header.Set("User-Agent", br.userAgent)
+	}
+
+	for key, value := range headers {
+		req.Header.Set(key, value)
+	}
+
+	rsp, err := br.doer.Do(req)
+	if err != nil {
+		return nil, ConnectionError{err}
+	}
+
+	return rsp, nil
+}
+
+// raw builds in a retry mechanism for GET failures (body-less request)
+func (br *BaseRequester) rawWithRetry(ctx context.Context, method, urlpath string, query url.Values, headers map[string]string, body io.Reader) (*http.Response, error) {
+	retry := time.NewTicker(rawRetry)
+	defer retry.Stop()
+	timeout := time.After(rawTimeout)
+	var rsp *http.Response
+	var err error
+	for {
+		rsp, err = br.raw(ctx, method, urlpath, query, headers, body)
+		if err == nil || method != "GET" {
+			break
+		}
+		select {
+		case <-retry.C:
+			continue
+		case <-timeout:
+		}
+		break
+	}
+	if err != nil {
+		return nil, err
+	}
+	return rsp, nil
+}
+
+func (br *BaseRequester) Do(ctx context.Context, opts *RequestOptions, result interface{}) (*RequestResponse, error) {
+	// Is the result expecting a websocket?
+	if ws, ok := result.(*Websocket); ok {
+		dialer := websocket.Dialer{
+			NetDial:          br.transport.Dial,
+			Proxy:            br.transport.Proxy,
+			TLSClientConfig:  br.transport.TLSClientConfig,
+			HandshakeTimeout: 5 * time.Second,
+		}
+		conn, _, err := dialer.DialContext(ctx, opts.Path, nil)
+		if err != nil {
+			return nil, err
+		}
+		*ws = conn
+		return nil, nil
+	}
+
+	// Is the result expecting a caller managed body reader?
+	if bodyReader, ok := result.(*BodyReader); ok {
+		httpResp, err := br.raw(ctx, opts.Method, opts.Path, opts.Query, opts.Headers, opts.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		*bodyReader = httpResp.Body
+		return nil, nil
+	}
+
+	// This is a normal sync or async server request
+	httpResp, err := br.rawWithRetry(ctx, opts.Method, opts.Path, opts.Query, opts.Headers, opts.Body)
+	if err != nil {
+		return nil, err
+	}
+	defer httpResp.Body.Close()
+
+	// Get the client decoder to extract what it needs before we proceed
+	reqResp, err := br.decoder(ctx, httpResp, opts, result)
+	if err != nil {
+		return nil, err
+	}
+
+	// Sanity check sync and async requests
+	if opts.Async == false {
+		if reqResp.Type != "sync" {
+			return nil, fmt.Errorf("expected sync response, got %q", reqResp.Type)
+		}
+	} else {
+		if reqResp.Type != "async" {
+			return nil, fmt.Errorf("expected async response for %q on %q, got %q", opts.Method, opts.Path, reqResp.Type)
```

Comment:
Why not be aligned with sync requests on this particular error?
```suggestion
			return nil, fmt.Errorf("expected async response, got %q", reqResp.Type)
```